### ISSUE TITLE
Extraction of Neumann boundary

### ISSUE THREAD

**Issue by DanielBoigk** [gridap/Gridap.jl#1008]
Hello all,

My Problem is: I want to calculate both dirichlet-to-neumann maps and neumann-to-dirichlet maps.
However I fail to reconstruct the values of the neumann boundary from the solution.
If you guys know how the problem gets the neumann boundary in the first place, maybe you can give the hint on how to extract a neumann boundary properly.

```julia
Ω = Triangulation(mesh)
dΩ = Measure(Ω,2)
Γ = BoundaryTriangulation(mesh, tags= "boundary" )
dΓ = Measure(Γ,2) 
# Weak problem:
a(u, v) = ∫( γ * ∇(v) ⋅ ∇(u) )dΩ
l(v) = ∫( b * v )dΓ  #here it gets the Neumann condition
```

Because sofar all my experiments on extracting it have failed. My input differs from my output.
Like this: 
![image](https://github.com/gridap/Gridap.jl/assets/71269893/889e65f9-e196-4a7e-84c0-a81aa5a5e34d)

**Comment by JordiManyer** [comment]
I would try something along the lines of 

```julia
uh = solve(...)
n = get_normal_vector(Γ)

u_N = ∇(uh)⋅n
```

which returns a `CellField` on `Γ` that you can evaluate on your boundary.

**Comment by DanielBoigk** [comment]
My solution was to assemble the Neumann problem with:

```julia
op =AffineFEOperator(a,l,U,V)
```

and then get the Matrix with 

```julia
K = op.op.matrix
uh = solve(op)
u = uh.free_values
```
Now i can extract the Neumann boundary was defined in the problem with:
```julia
f_out = K * u
``` 
if I select the entries corresponding to the boundary points I can get value that was put in out.

The Problem however was the following:
Even though the $\gamma$ function integrated to zero in the analytical sense it didn't sum up to zero on the points where the boundary was evaluated. Thus the system became overdetermined.
Is there a way to make enforce sum zero as an option in Gridap? While giving the boundary as a function is convenient it produces a faulty solution in this case, that need to be handled by explicitely editing:
```julia
f = op.op.vector
```
Many boundary value problems have the constraint that something sums up/integrates to zero.

**Comment by JordiManyer** [comment]
> Is there a way to make enforce sum zero as an option in Gridap?
Yes, it's the `:zerosum` constraint (`ZeroMeanFESpace`). It does not work for spaces with dirichlet BCs, however. On the other hand, you can use a lagrange multiplier approach using a `ConstantFESpace`.

**Comment by DanielBoigk** [comment]
@JordiManyer Another question:
So I have some minimzation functional:

$$ \min_{\gamma}J(\gamma,u) = \min_{\gamma}\sum\limits_{i=0}^n J_i(\gamma,u_i) $$

fulfilling some PDE:

$$ f(\gamma,u_i)=0 $$

with Neumann conditions:

$$ \frac{\partial u_i}{\partial \textbf{n}} = g_i $$

For a list of boundary conditions $g_1,...,g_n$ and functionals $J_1,...J_n$.
Since that functional is rather complicated and I haven't gotten Zygote to work with it I'm using adjoint methods to solve it.
I'm solving the state and the adjoint multiple times for all boundary conditions.

However now I need to calculate the functional derivative:

$$ \frac{\partial J}{\partial\gamma} = - \sum\limits_{i=1}^n \nabla u_i \cdot \nabla \lambda_i $$

which I evaluate in Gridap as:

`
grad_J[i] = - interpolate_everywhere( ∇(u[i]) ⋅ ∇(λ[i]) , V ).free_values
`

This happens to be very slow. It works, but is slower than solving an entire PDE multiple times.
**The Question is:**
For a given FESpace and map: Is there some way to assemble a tensor who calculates that bilinear map? Is there another way to achieve faster evaluation?

**Comment by zjwegert** [comment]
@DanielBoigk You may want to look at GridapTopOpt, specifically the [AD capabilities](https://zjwegert.github.io/GridapTopOpt.jl/dev/examples/AD-PDE-constrained-functions/). You might also find the [`RepeatingAffineFEStateMap`](https://zjwegert.github.io/GridapTopOpt.jl/dev/reference/statemaps/#RepeatingAffineFEStateMap) useful for having solutions to multiple FE problems with the same bilinear form
